plugins/inputs/snmp_trap/snmp_trap.go

```diff
@@ -0,0 +1,144 @@
+package snmp_trap
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/snmp"
+
+	"github.com/soniah/gosnmp"
+)
+
+type SnmpTrap struct {
+	Port uint16 `toml:"port"`
+
+	acc      telegraf.Accumulator
+	listener *gosnmp.TrapListener
+	wg       sync.WaitGroup
+	timeFunc func() time.Time
+
+	makeHandlerWrapper func(func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr)) func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr)
+	Errch              chan error
+}
+
+var sampleConfig = `
+  ## Port to listen on.  Default 162
+  #port = 162
+`
+
+func (s *SnmpTrap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *SnmpTrap) Description() string {
+	return "Receive SNMP traps"
+}
+
+func (s *SnmpTrap) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func init() {
+	inputs.Add("snmp_trap", func() telegraf.Input {
+		return &SnmpTrap{
+			timeFunc: time.Now,
+			Port:     162,
+		}
+	})
+}
+
+func (s *SnmpTrap) Init() error {
+	return nil
+}
+
+func (s *SnmpTrap) Start(acc telegraf.Accumulator) error {
+	s.acc = acc
+	s.listener = gosnmp.NewTrapListener()
+	s.listener.OnNewTrap = makeTrapHandler(s)
+	s.listener.Params = gosnmp.Default
+	//s.listener.Params.Logger = log.New(os.Stdout, "", 0)
+
+	//wrap the handler, used in unit tests
+	if nil != s.makeHandlerWrapper {
+		s.listener.OnNewTrap = s.makeHandlerWrapper(s.listener.OnNewTrap)
+	}
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		//no ip means listen on all interfaces, ipv4 and ipv6
+		err := s.listener.Listen(":" + strconv.FormatUint(uint64(s.Port), 10))
```

Comment:
I'll do this.  The test uses Listening() to make sure things are ready before sending a trap.  It will be nicer to move that code from the test into Start().  It will get rid of Errch too.